sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c

```diff
@@ -0,0 +1,1546 @@
+    fbuf->sfr.arq_timeout_event.event.next = NULL;
+    if (CONFIG_GNRC_SIXLOWPAN_SFR_INTER_FRAME_GAP_US > 0) {
+        for (clist_node_t *node = clist_lpop(&_frame_queue);
+             node != NULL; node = clist_lpop(&_frame_queue)) {
+            _frame_queue_t *entry = (_frame_queue_t *)node;
+            /* remove frames of this datagram from frame queue */
+            if (entry->datagram_tag == fbuf->tag) {
+                gnrc_pktbuf_release(entry->frame);
+                /* unset packet just to be safe */
+                entry->frame = NULL;
+                clist_rpush(&_frag_descs_free, node);
+            }
+            else {
+                clist_rpush(&new_queue, node);
+            }
+        }
+        /* reset frame queue with remaining frames */
+        _frame_queue = new_queue;
+    }
+    fbuf->offset = 0U;
+    fbuf->sfr.cur_seq = 0U;
+    fbuf->sfr.frags_sent = 0U;
+    fbuf->sfr.window_size = CONFIG_GNRC_SIXLOWPAN_SFR_OPT_WIN_SIZE;
+    for (clist_node_t *node = clist_lpop(&fbuf->sfr.window);
+         node != NULL; node = clist_lpop(&fbuf->sfr.window)) {
+        clist_rpush(&_frag_descs_free, node);
+    }
+}
+
+static gnrc_pktsnip_t *_build_rfrag(uint8_t tag, bool ack_req, uint16_t size,
+                                    uint8_t seq)
+{
+    sixlowpan_sfr_rfrag_t *hdr;
+    gnrc_pktsnip_t *frag = gnrc_pktbuf_add(NULL, NULL,
+                                           sizeof(sixlowpan_sfr_rfrag_t) +
+                                           size, GNRC_NETTYPE_SIXLOWPAN);
+
+    if (frag == NULL) {
+        return NULL;
+    }
+    sixlowpan_sfr_rfrag_set_disp(frag->data);
+
+    hdr = frag->data;
+    hdr->base.tag = tag;
+    if (ack_req) {
+        sixlowpan_sfr_rfrag_set_ack_req(hdr);
+    }
+    else {
+        sixlowpan_sfr_rfrag_clear_ack_req(hdr);
+    }
+    sixlowpan_sfr_rfrag_set_frag_size(hdr, size);
+    sixlowpan_sfr_rfrag_set_seq(hdr, seq);
+    /* set offset / datagram_size in callers */
+    return frag;
+}
+
+static gnrc_pktsnip_t *_build_frag_pkt(gnrc_netif_hdr_t *old_netif_hdr,
+                                       uint8_t tag, bool ack_req, uint16_t size,
+                                       uint8_t seq)
+{
+    gnrc_netif_hdr_t *new_netif_hdr;
+    gnrc_pktsnip_t *netif, *res;
+
+    DEBUG("6lo sfr: building fragment (tag: %u, X: %i, seq: %u, frag_size: %u)\n",
+           tag, ack_req, seq, size);
+    netif = gnrc_netif_hdr_build(gnrc_netif_hdr_get_src_addr(old_netif_hdr),
+                                 old_netif_hdr->src_l2addr_len,
+                                 gnrc_netif_hdr_get_dst_addr(old_netif_hdr),
+                                 old_netif_hdr->dst_l2addr_len);
+    if (netif == NULL) {
+        return NULL;
+    }
+
+    new_netif_hdr = netif->data;
+    *new_netif_hdr = *old_netif_hdr;
+    res = _build_rfrag(tag, ack_req, size, seq);
+    if (res == NULL) {
+        gnrc_pktbuf_release(netif);
+        return NULL;
+    }
+    LL_PREPEND(res, netif);
```

Comment:
Guess you could use the new gnrc_list functions here ;)

Reply:
Done.